Stripe/STPApplePayContext.m

```diff
@@ -0,0 +1,283 @@
+
+#import "STPAPIClient+ApplePay.h"
+#import "STPPaymentMethod.h"
+#import "STPPaymentIntentParams.h"
+#import "STPPaymentIntent+Private.h"
+#import "STPPaymentHandler.h"
+#import "NSError+Stripe.h"
+
+static char kSTPApplePayContextAssociatedObjectKey;
+
+typedef NS_ENUM(NSUInteger, STPPaymentState) {
+    STPPaymentStateNotStarted,
+    STPPaymentStatePending,
+    STPPaymentStateError,
+    STPPaymentStateSuccess
+};
+
+@interface STPApplePayContext() <PKPaymentAuthorizationViewControllerDelegate>
+
+@property (nonatomic, weak) id<STPApplePayContextDelegate> delegate;
+@property (nonatomic) PKPaymentAuthorizationViewController *viewController;
+
+// Internal state
+@property (nonatomic) STPPaymentState paymentState;
+@property (nonatomic, nullable) NSError *error;
+/// YES if the flow cancelled or timed out.  This toggles which delegate method (didFinish or didiAuthorize) calls our didComplete delegate method
+@property (nonatomic) BOOL didCancelOrTimeoutWhilePending;
+@property (nonatomic) BOOL didPresentApplePay;
+
+@end
+
+@implementation STPApplePayContext
+
+- (instancetype)initWithPaymentRequest:(PKPaymentRequest *)paymentRequest delegate:(id<STPApplePayContextDelegate>)delegate {
+    if (![Stripe canSubmitPaymentRequest:paymentRequest]) {
+        return nil;
+    }
+    
+    self = [super init];
+    if (self) {
+        _paymentState = STPPaymentStateNotStarted;
+        _delegate = delegate;
+        _apiClient = [STPAPIClient sharedClient];
+        _viewController = [[PKPaymentAuthorizationViewController alloc] initWithPaymentRequest:paymentRequest];
+        _viewController.delegate = self;
+        
+        if (_viewController == nil) {
+            return nil;
+        }
+    }
+    return self;
+}
+
+- (void)setApiClient:(STPAPIClient *)apiClient {
+    if (apiClient == nil) {
+        _apiClient = [STPAPIClient sharedClient];
+    } else {
+        _apiClient = apiClient;
+    }
+}
+
+- (void)presentApplePayOnViewController:(UIViewController *)viewController completion:(STPVoidBlock)completion {
+    if (self.didPresentApplePay) {
+        NSAssert(NO, @"This method should only be called once; create a new instance every time you present Apple Pay.");
+        return;
+    }
+    self.didPresentApplePay = YES;
+    
+    // This instance must live so that the apple pay sheet is dismissed; until then, the app is effectively frozen.
+    objc_setAssociatedObject(self.viewController, &kSTPApplePayContextAssociatedObjectKey, self, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+
+    [viewController presentViewController:self.viewController animated:YES completion:completion];
+}
+
+- (void)forwardInvocation:(NSInvocation *)invocation {
+    // Called for the methods we additionally respond YES to in `respondsToSelector`, letting us forward directly to self.delegate
+    // We could alternatively implement the PKPaymentAuthorizationViewControllerDelegate methods to call their respective STPApplePayContextDelegate methods
```

Comment:
hmm, I think I'd actually prefer this approach. It seems less complicated and would also respond to delegate method changes better (e.g. if one gets marked as deprecated, etc.)